pkg/proxy/apis/config/types.go

```diff
@@ -89,6 +89,18 @@ type KubeProxyConntrackConfiguration struct {
 	// in CLOSE_WAIT state will remain in the conntrack
 	// table. (e.g. '60s'). Must be greater than 0 to set.
 	TCPCloseWaitTimeout *metav1.Duration
+	// tcpBeLiberal is the value kube-proxy will set for
+	// nf_conntrack_tcp_be_liberal sysctl
+	// (e.g. 1). Must be a non-negative integer to set.
+	TCPBeLiberal *int32
```

Comment:
this is a boolean, why making this an integer?

```
nf_conntrack_tcp_be_liberal - BOOLEAN
	0 - disabled (default)
	not 0 - enabled

	Be conservative in what you do, be liberal in what you accept from others.
	If it's non-zero, we mark only out of window RST segments as INVALID.
```
https://www.kernel.org/doc/Documentation/networking/nf_conntrack-sysctl.txt

Reply:
I must have got confused with "not 0" part, just wondering why they didn't used "1" instead of "not 0" :p
Will fix this 🙂

Reply:
lol

Reply:
@aojea I tried to use bool here but [BoolVar](https://pkg.go.dev/github.com/spf13/pflag#BoolVar) takes a default value of type **bool** not ***bool**, we would have to use a default value, and can't keep it as nil. Here the default behaviour should be not to update the **nf_conntrack_tcp_be_liberal** variable. 

I believe that we need three states for this:
1. On/Enabled/True/1      (set nf_conntrack_tcp_be_liberal = 1)
2. Off/Disabled/False/0  (set nf_conntrack_tcp_be_liberal = 0)
3. nil/-1 [keep as it is]     (leave it as it is)

LMK if there is any way around for this using boolean or I can use int32 only with an additional validation of values to be either **0** or **1**.

Reply:
omg, the config vs flag thing again, https://github.com/kubernetes/kubernetes/issues/108737#issuecomment-1688079964

Reply:
there's a `cliflags.Tristate` but I don't think it serializes to a `*bool`, so you'd need to have the config option be one thing and read the command-line flag into a variable and then sync the two manually like we do with some other "weird" options

Reply:
yes `cliflags.Tristate` serializes to **bool** only and we can't directly serialize it to ***bool** as [receiver base type cannot be a pointer](https://go.dev/ref/spec#Method_declarations:~:text=A%20receiver%20base%20type%20cannot%20be%20a%20pointer%20or%20interface%20type) in go.

> so you'd need to have the config option be one thing and read the command-line flag into a variable and then sync the two manually

So should I use ***bool** in config and add a new type `TriStateVar` in `pkg/util/flag` which implements **pflag.Value** interface to sync values as we do for **healthz-bind-address** and **metrics-bind-address** ?

```
type TriStateVar struct {
	Val *bool
}
```

Reply:
but why do you need to serialize it to bool, Tristate can identify if the value was set or not

Reply:
@aojea @danwinship I'm using `*bool` in config and `cliflags.Tristate` in command-line flag. 
Update the code.

Reply:
@aroradaman what Dan says https://github.com/kubernetes/kubernetes/pull/120354#issuecomment-1709015636 😄